src/storage/checkpoints/checkpoint.cpp

```diff
@@ -0,0 +1,144 @@
+
+
+#include "storage/checkpoints/checkpoint.h"
+#include <common/worker_pool.h>
+#include <fstream>
+namespace terrier::storage {
+
+bool Checkpoint::TakeCheckpoint(const std::string &path, catalog::db_oid_t db) {
+  // get db catalog accessor
+  auto accessor = catalog_->GetAccessor(txn_, db);
+  std::unordered_set<catalog::table_oid_t> table_oids = accessor->GetAllTableOids();
+
+  for (const auto &oid : table_oids) {
+    queue.push_back(oid);
+  }
+
+  // initalize threads
+  auto num_threads = 1u;
+  common::WorkerPool thread_pool_{num_threads, {}};
+  thread_pool_.Startup();
+  auto workload = [&](uint32_t worker_id) {
+    // copy contents of table to disk
+    WriteToDisk(path, accessor, db);
+  };
+  for (auto i = 0u; i < num_threads; i++) {
+    thread_pool_.SubmitTask([i, &workload] { workload(i); });
+  }
+  thread_pool_.WaitUntilAllFinished();
+  if (queue.size() > 0) {
+    // the table oid that failed to be backup
+    return false;
+  }
+  return true;
+}
+
+void Checkpoint::WriteToDisk(const std::string &path, const std::unique_ptr<catalog::CatalogAccessor> &accessor,
+                             catalog::db_oid_t db_oid) {
+  while (queue.size() > 0) {
+    queue_latch.lock();
+    if (queue.size() <= 0) {
+      queue_latch.unlock();
+      return;
+    }
+    catalog::table_oid_t curr_table_oid = queue[0];
```

Comment:
While pulling a table oid to checkpoint from a queue whenever a thread is free, is certainly a good approach, I feel a better approach (if you are sticking with a thread handling writetodisk for multiple tables) is to sort the list of tables according to the size and then partitioning them for each thread. This way, you also avoid having a mutex for the queue as the partitioning is already done for each thread. I do realize there's a chance that a table oid is pushed back to the queue, but to handle that you can have multiple iterations of partitioning and checkpointing. 

A simple example of how this strategy can be beneficial: 
Say there are 4 tables, table sizes are:  1G, 1G, 1G, 10G. Let's say time taken to write 1G = 10s. 
Say you had 3 threads/cores. According to the current logic, time taken to write to disk = 110s (the last thread handles 11G). According to the partitioning logic, the first thread handles 10G, second thread handles 2G, third one handles 1G. Time taken = 100s. 

I'm not very sure what exactly is the design that you would go with in future, but this is just a thought that you may want to consider.